src/convolution.jl

```diff
@@ -0,0 +1,100 @@
+"""
+    convolve(d1::T, d2::T) where T<:Distribution -> Distribution
+
+Convolve two distributions of the same type to yield the distribution corresponding to the
+sum of independent random variables drawn from the underlying distributions.
+
+The function is only defined in the cases where the convolution has a closed form as
+defined here https://en.wikipedia.org/wiki/List_of_convolutions_of_probability_distributions
+
+* Bernoulli
+* Binomial
+* Negative Binomial
+* Geometric
+* Poisson
+* Normal
+* Cauchy
+* Chisq
+* Exponential
+* Gamma
+* Multi-variate Normal
+"""
+#convolve(d1, d2) = throw(ArgumentError("`convolve` is not defined for $(typeof(d1)) and $(typeof(d2))"))
+
+# discrete univariate
+function convolve(d1::Bernoulli, d2::Bernoulli)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(2, d1.p)
+end
+
+function convolve(d1::Binomial, d2::Binomial)
+    _check_convolution_args(d1.p, d2.p)
+    return Binomial(d1.n + d2.n, d1.p)
+end
+
+function convolve(d1::NegativeBinomial, d2::NegativeBinomial)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(d1.r + d2.r, d1.p)
+end
+
+function convolve(d1::Geometric, d2::Geometric)
+    _check_convolution_args(d1.p, d2.p)
+    return NegativeBinomial(2, d1.p)
+end
+
+convolve(d1::Poisson, d2::Poisson) =  Poisson(d1.λ + d2.λ)
+
+
+# continuous univariate
+convolve(d1::Normal, d2::Normal) = Normal(d1.μ + d2.μ, hypot(d1.σ, d2.σ))
+convolve(d1::Cauchy{T}, d2::Cauchy{T}) where T = Cauchy(d1.μ + d2.μ, d1.σ + d2.σ)
+convolve(d1::Chisq, d2::Chisq) = Chisq(d1.ν + d2.ν)
+
+function convolve(d1::Exponential, d2::Exponential)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(2, d1.θ)
+end
+
+function convolve(d1::Gamma, d2::Gamma)
+    _check_convolution_args(d1.θ, d2.θ)
+    return Gamma(d1.α + d2.α, d1.θ)
+end
+
+# continuous multivariate
+function convolve(
+    d1::Union{IsoNormal, ZeroMeanIsoNormal},
+    d2::Union{IsoNormal, ZeroMeanIsoNormal},
+    )
+    _check_convolution_shape(d1, d2)
+    return MvNormal(d1.μ .+ d2.μ, d1.Σ.value + d2.Σ.value)
```

Comment:
we don't need the `.value` here

```julia
julia> z1 = MvNormal(Distributions.ZeroVector(Float64, 100), I)
ZeroMeanIsoNormal(
dim: 100
μ: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> z2 = MvNormal(ones(100), I)
IsoNormal(
dim: 100
μ: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0  …  1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> @btime MvNormal($z1.μ .+ $z2.μ, $z1.Σ + $z2.Σ);
  87.319 ns (2 allocations: 944 bytes)

julia> @btime MvNormal($z1.μ .+ $z2.μ, $z1.Σ.value + $z2.Σ.value);
  87.682 ns (2 allocations: 944 bytes)
```

Reply:
actually, the `*DiagNormal` and `*IsoNormal` can be combined
```julia
julia> z1 = MvNormal(repeat([1.0], 100))
ZeroMeanDiagNormal(
dim: 100
μ: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> z2 = MvNormal(ones(100), repeat([1.0], 100))
DiagNormal(
dim: 100
μ: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0  …  1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> z3 = MvNormal(Distributions.ZeroVector(Float64, 100), I)
ZeroMeanIsoNormal(
dim: 100
μ: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> z4 = MvNormal(ones(100), I)
IsoNormal(
dim: 100
μ: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0  …  1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
Σ: [1.0 0.0 … 0.0 0.0; 0.0 1.0 … 0.0 0.0; … ; 0.0 0.0 … 1.0 0.0; 0.0 0.0 … 0.0 1.0]
)


julia> for (a, b) in Iterators.product((z1, z2), (z3, z4))
           @btime MvNormal($a.μ .+ $b.μ, $a.Σ + $b.Σ)
       end
  271.267 ns (5 allocations: 2.69 KiB)
  264.330 ns (5 allocations: 2.69 KiB)
  260.786 ns (5 allocations: 2.69 KiB)
  263.312 ns (5 allocations: 2.69 KiB)
```

Reply:
(which makes since, since they both have diagonal cov matrices)

Reply:
nice, that simplifies things a bit

Reply:
i guess we can use a const to save us a bit here
```julia
const IsoOrDiagNormal = Union{IsoNormal, ZeroMeanIsoNormal, DiagNormal, ZeroMeanDiagNormal}`

function convolve(d1::IsoOrDiagNormal, d2::IsoOrDiagNormal)
    check_convolution_shape(d1, d2)
    return MvNormal(d1.μ .+ d2.μ, d1.Σ + d2.Σ)
end
```